core/domain/translation_domain.py

```diff
@@ -18,9 +18,258 @@
+from typing import Dict, List
+from typing_extensions import TypedDict
+
+
+TRANSLATABLE_CONTENT_FORMAT_HTML = 'html'
+TRANSLATABLE_CONTENT_FORMAT_UNICODE_STRING = 'unicode'
+TRANSLATABLE_CONTENT_FORMAT_SET_OF_NORMALIZED_STRING = (
+    'set_of_normalized_string')
+TRANSLATABLE_CONTENT_FORMAT_SET_OF_UNICODE_STRING = 'set_of_unicode_string'
+
+
+class TranslatableContentDict(TypedDict):
+    """Dictionary representing TranslatableContent object."""
+
+    content_id: str
+    content: str|List[str]
+    content_type: str
+
+
+class TranslatedContentDict(TypedDict):
+    """Dictionary representing TranslatedContent object."""
+
+    content: str|List[str]
+    needs_update: bool
+
+
+class BaseTranslatableObject:
+    """Base class for all translatable objects."""
+
+    # Private field to track translatable contents in a BaseTranslatableObject.
+    _translatable_contents: Dict[str, TranslatableContent] = {}
+
+    def _register_all_translatable_fields(self) -> None:
+        """Method to register all translatable fields in a translatable object.
+
+        Raises:
+            NotImplementedError. The derived child class must implement the
+                necessary logic to register all translatable fields in an
+                entity.
+        """
+        raise NotImplementedError('Must be implemented in subclasses.')
+
+    def _register_translatable_field(
+        self,
+        field_type: str,
+        content_id: str,
+        value: str|List[str]
+    ) -> None:
+        """Method to register a translatable field in a translatable object.
+
+        Args:
+            field_type: str. The type of the corresponding translatable content.
+            content_id: str. The id of the corresponding translatable content.
+            value: str|list(str). Value of the content which is translatable.
+        """
+        if content_id in self._translatable_contents:
+            raise Exception(
+                'A translatable field is already registered with the '
+                'same content id: %s' % content_id)
+
+        self._translatable_contents[content_id] = TranslatableContent(
+            content_id, value, field_type)
+
+    def _register_translatable_object(
+        self,
+        value: BaseTranslatableObject
+    ) -> None:
+        """Method to register translatable field in a translatable object.
+
+        Args:
+            value: BaseTranslatableObject. An object representing
+                BaseTranslatableObject.
+        """
+        self._translatable_contents.update(value.get_translatable_fields())
+
+    def get_translatable_fields(self) -> Dict[str, TranslatableContent]:
+        """Method to get all the translatable fields in a translatable object.
+
+        Returns:
+            Dict(str, TranslatableContent). Returns the dict containg content_id
+            as key and TranslatableContent as value.
+        """
+        self._translatable_contents = {}
+        self._register_all_translatable_fields()
```

Comment:
Why do we need to call this on every GET operation? Is it possible for the list of translatable fields to change in the lifecycle of an object? 

If we keep this in, it may be worth leaving a comment to explain why it's needed.

Reply:
Yes, the state of translatable fields changes in the life cycle of an object.
We generally update or change a state of a field very often. Hence this method is designed in a way that at any moment if it is called, it will always return the latest state of an object.
Example: A test where we directly delete the hints list – [link](https://github.com/oppia/oppia/blob/34a5a5b4a6aea675d58fd48337643d554dfa93e4/core/domain/state_domain_test.py#L3601)

Reply:
Answer to question 1
- In order to avoid duplicates while registering a translatable field we keep a class level variable. If we make it local for get_translatable_fields() method then we have to pass an extra argument to _register_translatable_field() for checking duplicates.
- In below figure checking of duplicate contents for object Y can be easy, by class level variable otherwise content_id should be passed as argument in  _register_translatable_field() method.
In future if we add any other functionality to BaseTranslatableObject then it would be easy to access all translatable contents via _translatable_contents.
![image](https://user-images.githubusercontent.com/57531197/153723811-692c78e2-aee1-420c-9238-fb77f85b2eb4.png)

Reply:
Second answer: 
Currently _get_all_translatable_content() is used in our codebase and all the translatable_items are collected at one call, but this method is complex and based on the design doc we want to make every object independent with a view of translation.
Currently this method takes around ~225ms
![image](https://user-images.githubusercontent.com/57531197/153723863-b9cc07e6-6a56-4a9f-91db-1005182fb725.png)

Currently _get_all_translatable_content() method is called by these methods–
- get_translation_counts()
- get_translatable_content_count()
- has_content_id()
- get_content_id_mapping_needing_translations()

Sorry if I misinterpreted the question, looking to your suggestion.

Reply:
Thanks @Nik-09, here are my follow-ups. 

1. Do we ever delete items from the registration dict? If not, then how do you handle the "hint gets deleted" case?
2. I don't follow the argument about duplicates. In a composite structure, why would we have duplicates in the first place -- wouldn't that already indicate that something is wrong? What's a legitimate case in which we have duplicates?
3. "In future if we add any other functionality to BaseTranslatableObject then it would be easy to access all translatable contents via _translatable_contents." -- The problem with this is that it means you have to automatically keep _translatable_contents up-to-date any time a change is made to the object. How are you planning to do that? The two possible ways are to (a) recompute on each get, or (b) update the inner field on each update; I think you're going with (a), but that means that translatable content should always be accessed via the public get() method -- is that right? In that case I think it's OK to keep the dict as local within the get() method so that it doesn't get used outside its intended context.
4. How many *times* does _get_all_translatable_content() get called within a single request/response (typically)? Is it once for each of the bullet points you mentioned?

Reply:
Modified the architecture after discussion.

Reply:
Answered below: https://github.com/oppia/oppia/pull/14515#discussion_r818383394